book/src/proposals/validator-timestamp-oracle.md

```diff
@@ -0,0 +1,110 @@
+too much transaction overhead to the cluster. Validators can convert this time
+to a slot interval using the `slots_per_year` value that is stored in each bank.
+
+```text
+let seconds_in_30min = 1800;
+let timestamp_interval = (slots_per_year / SECONDS_PER_YEAR) * seconds_in_30min;
+```
+
+This transaction should be triggered in replay_stage (likely near where
+`subscriptions.notify_slot()` is executed) when `current_slot %
+timestamp_interval == 0`.
+
+### Timestamp Accounts
+
+Before beginning operation, a validator needs to create and initialize a
+Timestamp account that will hold its most recent slot-timestamp. Like the
+validator's vote account, this account needs to be rent exempt.
+
+### Timestamp Program
+
+The on-chain Timestamp program processes Timestamp transactions from validators.
+It needs to load the correct Timestamp account, verify that the transaction
+signer is the expected validator, compare the slot and timestamp to the
+currently stored values to verify that they are both monotonically increasing,
+and store the new slot and timestamp in the account.
+
+## Calculating Median Timestamp
+
+In order to calculate the estimated timestamp for a particular block, a
+validator first needs to identify the most recently timestamped slot:
+
+```text
+let timestamp_slot = floor(current_slot / timestamp_interval);
+```
+
+Then the validator needs to gather all Timestamp transactions from the ledger
+that reference that slot, using `Blocktree::get_slot_entries()`. As these
+transactions could have taken some time to reach and be processed by the leader,
+the validator needs to scan several completed blocks after the timestamp_slot to
+get a reasonable set of Timestamps. The exact number of slots will need to be
+tuned: More slots will enable greater cluster participation and more timestamp
+datapoints; fewer slots will speed how long Timestamp filtering takes.
+
+From this collection of transactions, the validator selects the median timestamp.
+
+Any validator replaying the ledger should derive the same median timestamp by processing the Timestamp transactions from the same number of slots.
+
+## Calculating Estimated Time for a Particular Block
+
+Once the median timestamp for a known slot is calculated, it is trivial to
+calculate the estimated timestamp for subsequent block N:
+
+```text
+let block_n_timestamp = median_timestamp + (block_n_slot_offset * slot_duration);
+```
+
+where `block_n_slot_offset` is the difference between the slot of block N and
+the timestamp_slot, and `slot_duration` is derived from the cluster's
+`slots_per_year` stored in each Bank
+
+## Additional Considerations
+
+1. This proposal puts a distinct processing burden on a validator when it receives
+a `getBlockTime` RPC request. Another approach would be for a validator to
+maintain a list of median timestamps, populated as the validator
+builds/replays the ledger.
+
+2. This proposal doesn't address the 4th requirement, that each validator must
```

Comment:
We could tie this into the uptime measurement somehow, where uptime is redefined as a function of a validator both promptly submitting votes and timestamps over an epoch.  While this wouldn't reward a validator for submitting timestamps, it would socially penalize validators that modify the code to disable timestamp submission and maybe that's enough